File: app/src/main/java/info/nightscout/androidaps/queue/commands/Command.java
package info.nightscout.androidaps.queue.commands;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.data.PumpEnactResult;
import info.nightscout.androidaps.logging.L;
import info.nightscout.androidaps.queue.Callback;

/**
 * Created by mike on 09.11.2017.
 */
public abstract class Command {
    private Logger log = LoggerFactory.getLogger(L.PUMPQUEUE);

    public enum CommandType {
        BOLUS,
        SMB_BOLUS,
        CARBS_ONLY_TREATMENT,
        TEMPBASAL,
        EXTENDEDBOLUS,
        BASALPROFILE,
        READSTATUS,
        LOADHISTORY, // TDDs and so far only Dana specific
        LOADEVENTS, // so far only Dana specific
        SETUSERSETTINGS, // so far only Dana specific,
        START_PUMP,
        STOP_PUMP,
        INSIGHT_SET_TBR_OVER_ALARM
    }

    public CommandType commandType;
    protected Callback callback;

    public abstract void execute();

    public abstract String status();

    public void cancel() {
        PumpEnactResult result = new PumpEnactResult();
        result.success = false;
        result.comment = MainApp.gs(R.string.connectiontimedout);
        if (L.isEnabled(L.PUMPQUEUE))
            log.debug("Result cancel");
        if (callback != null)
            callback.result(result).run();
    }
}

File: core/src/main/java/info/nightscout/androidaps/queue/commands/Command.kt
package info.nightscout.androidaps.queue.commands

import dagger.android.HasAndroidInjector
import info.nightscout.androidaps.core.R
import info.nightscout.androidaps.data.PumpEnactResult
import info.nightscout.androidaps.database.AppRepository
import info.nightscout.shared.logging.AAPSLogger
import info.nightscout.shared.logging.LTag
import info.nightscout.androidaps.queue.Callback
import info.nightscout.androidaps.utils.resources.ResourceHelper
import javax.inject.Inject

abstract class Command(
    val injector: HasAndroidInjector,
    val commandType: CommandType,
    val callback: Callback? = null
) {

    @Inject lateinit var aapsLogger: AAPSLogger
    @Inject lateinit var rh: ResourceHelper
    @Inject lateinit var repository: AppRepository

    enum class CommandType {
        BOLUS,
        SMB_BOLUS,
        CARBS_ONLY_TREATMENT,
        TEMPBASAL,
        EXTENDEDBOLUS,
        BASAL_PROFILE,
        READSTATUS,
        LOAD_HISTORY,  // TDDs and so far only Dana specific
        LOAD_EVENTS,  // so far only Dana specific
        LOAD_TDD,
        SET_USER_SETTINGS,  // so far only Dana specific,
        START_PUMP,
        STOP_PUMP,
        INSIGHT_SET_TBR_OVER_ALARM, // insight only
        CUSTOM_COMMAND
    }

    init {
        @Suppress("LeakingThis")
        injector.androidInjector().inject(this)
    }

    abstract fun execute()
    abstract fun status(): String
    abstract fun log(): String

    fun cancel() {
        val result = PumpEnactResult(injector)
        result.success = false
        result.comment = rh.gs(R.string.connectiontimedout)
        aapsLogger.debug(LTag.PUMPQUEUE, "Result cancel")
        callback?.result(result)?.run()
    }
}